File: cpp/stack.cpp
#include <stdio.h>
#include <iostream>

using namespace std;

struct Stack
{
private:
  unsigned capacity;
  int *items;
  unsigned last_index = 0;
  unsigned size = 0;

public:
  Stack(unsigned int c = 4) : capacity(c)
  {
    this->items = new int[c];
  }

  void clear()
  {
    while (this->size != 0)
    {
      this->items[this->size - 1] = NULL;
    }
  }

  bool empty()
  {
    return this->size == 0;
  }

  bool full()
  {
    return this->size == this->capacity;
  }

  void insert(int item)
  {
    if (this->full())
    {
      throw runtime_error("[Stack]: is full!");
    }

    this->last_index++;
    this->items[this->last_index] = item;
  }

  int pop()
  {
    int last_item = this->items[this->last_index];

    this->items[this->last_index] = NULL;
    this->last_index--;

    return last_index;
  }

  void show()
  {
    int i = this->size - 1;

    while (i != 0)
    {
      cout << this->items[i] << endl;
      i--;
    }
  }

  int size()
  {
    return this->size;
  }

  int top()
  {
    return this->items[this->last_index];
  }
};

File: rust/src/stack.rs
use std::io::{Error, ErrorKind};

struct Stack {
    capacity: u64,
    items: Vec<Option<i64>>,
    last_index: i64,
}

impl Stack {
    pub fn new(capacity: u64) -> Self {
        Self {
            capacity,
            items: Vec::with_capacity(capacity as usize),
            last_index: -1,
        }
    }

    pub fn clear(&mut self) {
        while self.last_index > -1 {
            self.items[self.last_index as usize] = None;
            self.last_index -= 1;
        }
    }

    pub fn empty(&self) -> bool {
        self.last_index == -1
    }

    pub fn full(&self) -> bool {
        self.size() == self.capacity
    }

    pub fn insert(&mut self, item: i64) -> Result<(), Error> {
        if self.full() {
            return Err(Error::new(ErrorKind::Other, "[stack] is full!"));
        }

        self.last_index += 1;
        self.items[self.last_index as usize] = Some(item);

        Ok(())
    }

    pub fn pop(&mut self) {
        if self.empty() {
            return;
        }

        self.items[self.last_index as usize] = None;
        self.last_index -= 1
    }

    pub fn size(&self) -> u64 {
        (self.last_index + 1) as u64
    }

    pub fn show(&self) {
        let mut i = self.last_index as usize;

        while i != 0 {
            println!("{}", self.items[i].unwrap());
            i -= 1
        }
    }

    pub fn top(&self) -> Option<i64> {
        self.items[self.last_index as usize]
    }
}
